### PR TITLE
Create JsonRpcSigningServer

### LINKED ISSUE #801: Create JsonRpcSigningServer wrapper around JsSigningServer
Requires #800

This does the same as JsSigningServer but adds a layer of JSON encoding/decoding

### LINKED ISSUE #914: Export isJsRpcErrorResponse from @iov/core

### PR DESCRIPTION
**Pull request by webmaster128**
Closes #801
Closes #914

---

JsonRpcSigningServer is like JsRpcSigningServer but works over every JSON-based channel. Unfortunately the [website browser extension communication](https://github.com/iov-one/ponferrada/wiki/Communication-channels) does not support the [structured clone algorithm](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm) (website to WebWorker does), so we need a way to serialize/deserialize Uint8Array as JSON, which is done using "bytes:" prefixed hex strings.

Instead of wrapping JsonRpcSigningServer around JsRpcSigningServer, which adds mostly unnecessary complexity, I changed my mind and copied most of JsRpcSigningServer to prepare the removal of JsRpcSigningServer.

### CODE REVIEW

File: packages/iov-core/src/jsonrpcsigningserver.ts [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        },
+      };
+      return errorResponse;
+    }
+
+    return this.handleChecked(checkedRequest);
+  }
+
+  /**
+   * Handles a checked JsonRpcRequest
```
Comment (willclarktech) [inline packages/iov-core/src/jsonrpcsigningserver.ts]: What does "checked" mean exactly?

File: packages/iov-core/src/jsonrpcsigningserver.ts [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+  public async handleChecked(request: JsonRpcRequest): Promise<JsonRpcResponse> {
+    let call: RpcCall;
+    try {
+      call = parseRpcCall(request);
+    } catch (error) {
+      const errorResponse: JsonRpcErrorResponse = {
+        jsonrpc: "2.0",
+        id: request.id,
+        error: {
+          code: jsonRpcCode.methodNotFound,
```
Comment (willclarktech) [inline packages/iov-core/src/jsonrpcsigningserver.ts]: What if the error is a `ParamsError`?

File: packages/iov-core/src/signingservice.spec.ts [hunk 3]
```diff
@@ -137,13 +139,14 @@ describe("signingservice.worker", () => {
...
-      fromHex("533e376559fa551130e721735af5e7c9fcd8869ddd54519ee779fce5984d7898"),
-    );
+    const result = TransactionEncoder.fromJson(response.result);
+    expect(result).toEqual(jasmine.any(Array));
+    expect((result as ReadonlyArray<any>).length).toEqual(1);
+    expect(result[0].chainId).toEqual(bnsConnection.chainId());
+    expect(result[0].pubkey).toEqual({
+      algo: Algorithm.Ed25519,
+      data: fromHex("533e376559fa551130e721735af5e7c9fcd8869ddd54519ee779fce5984d7898"),
+    });
```
Comment (willclarktech) [inline packages/iov-core/src/signingservice.spec.ts]: As above re object comparison.

File: packages/iov-core/src/transactionencoder.spec.ts [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+    it("works for dicts", () => {
+      expect(toJson({})).toEqual({});
+      expect(toJson({ foo: 1 })).toEqual({ foo: 1 });
+      expect(toJson({ foo: 1, bar: 2 })).toEqual({ foo: 1, bar: 2 });
+      expect(toJson({ foo: { bar: 1 } })).toEqual({ foo: { bar: 1 } });
+    });
+
+    it("fails for unsupported objects", () => {
+      expect(() => toJson(() => 0)).toThrowError(/Cannot encode type to JSON/i);
```
Comment (willclarktech) [inline packages/iov-core/src/transactionencoder.spec.ts]: How about storing the regex in a variable so it's clear they all fail with the same message?

File: packages/iov-core/src/transactionencoder.spec.ts [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+      expect(() => toJson(/[0-9]+/)).toThrowError(/Cannot encode type to JSON/i);
+      expect(() => toJson(new Uint32Array())).toThrowError(/Cannot encode type to JSON/i);
+      expect(() => toJson(undefined)).toThrowError(/Cannot encode type to JSON/i);
+    });
+
+    it("prefixes strings", () => {
+      expect(toJson("")).toEqual("string:");
+      expect(toJson("abc")).toEqual("string:abc");
+      expect(toJson("\0")).toEqual("string:\0");
```
Comment (willclarktech) [inline packages/iov-core/src/transactionencoder.spec.ts]: What about a string which starts `string:`?

File: packages/iov-core/src/transactionencoder.spec.ts [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+      expect(fromJson("string:abc")).toEqual("abc");
+      expect(fromJson("string:\0")).toEqual("\0");
+
+      // "bytes:" prefix
+      expect(fromJson("bytes:")).toEqual(new Uint8Array([]));
+      expect(fromJson("bytes:12")).toEqual(new Uint8Array([0x12]));
+      expect(fromJson("bytes:aabb")).toEqual(new Uint8Array([0xaa, 0xbb]));
+
+      // other prefixes
+      expect(() => fromJson("")).toThrowError(/Found string with unknown prefix/i);
```
Comment (willclarktech) [inline packages/iov-core/src/transactionencoder.spec.ts]: As above re regex.

File: packages/iov-core/src/transactionencoder.spec.ts [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    });
+  });
+
+  describe("fromJson", () => {
+    it("works for numbers", () => {
+      expect(fromJson(0)).toEqual(0);
+      expect(fromJson(1)).toEqual(1);
+      expect(fromJson(-1)).toEqual(-1);
+      expect(fromJson(Number.POSITIVE_INFINITY)).toEqual(Number.POSITIVE_INFINITY);
+      expect(fromJson(Number.NaN)).toEqual(Number.NaN);
```
Comment (willclarktech) [inline packages/iov-core/src/transactionencoder.spec.ts]: Numbers beyond safe range?

File: packages/iov-core/src/transactionencoder.spec.ts [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+
+    it("prefixes and hex encodes Uint8Array", () => {
+      expect(toJson(new Uint8Array([]))).toEqual("bytes:");
+      expect(toJson(new Uint8Array([0x12]))).toEqual("bytes:12");
+      expect(toJson(new Uint8Array([0x12, 0xab]))).toEqual("bytes:12ab");
+    });
+
+    // Encoding recursive objects is explicitly undefined behavior. Just use this
+    // test to play around.
+    xit("fails for recursive objects", () => {
```
Comment (willclarktech) [inline packages/iov-core/src/transactionencoder.spec.ts]: Does this explicitly undefined behaviour also cover recursive arrays?

File: packages/iov-core/src/transactionencoder.ts [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    if (Array.isArray(data)) {
+      return data.map(TransactionEncoder.toJson);
+    }
+
+    // Exclude special kind of objects like Array, Date or Uint8Array
+    // Object.prototype.toString() returns a specified value:
+    // http://www.ecma-international.org/ecma-262/7.0/index.html#sec-object.prototype.tostring
+    if (
+      typeof data === "object" &&
+      data !== null &&
```
Comment (willclarktech) [inline packages/iov-core/src/transactionencoder.ts]: `null` check not strictly necessary because of above `return`.

File: packages/iov-core/src/transactionencoder.ts [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+      Object.prototype.toString.call(data) === "[object Object]"
+    ) {
+      const out: any = {};
+      for (const key of Object.keys(data)) {
+        // tslint:disable-next-line: no-object-mutation
+        out[key] = TransactionEncoder.toJson((data as any)[key]);
```
Comment (willclarktech) [inline packages/iov-core/src/transactionencoder.ts]: `data as Object`?

File: packages/iov-core/src/transactionencoder.ts [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+      return data;
+    }
+
+    if (data === null) {
+      return null;
+    }
+
+    if (typeof data === "string") {
+      if (data.startsWith("string:")) {
+        return data.slice(7);
```
Comment (willclarktech) [inline packages/iov-core/src/transactionencoder.ts]: Prefer a solution without magic numbers.

File: packages/iov-core/src/jsonrpcsigningserver.spec.ts [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+      algo: Algorithm.Ed25519,
+      data: fromHex("533e376559fa551130e721735af5e7c9fcd8869ddd54519ee779fce5984d7898"),
+    });
+    expect(result[1]).toEqual(ganacheSecondIdentity);
+
+    server.shutdown();
+    bnsConnection.disconnect();
+  });
+
+  it("send a signing request to service", async () => {
```
[the remark this answers is not in the document]
Comment (webmaster128) [reply packages/iov-core/src/jsonrpcsigningserver.spec.ts]: Oh, even worse. The unit under test we describe here does not _send_ but _handle_.

File: packages/iov-core/src/jsonrpcsigningserver.spec.ts [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+      params: {
+        reason: "Please sign",
+        transaction: TransactionEncoder.toJson(send),
+      },
+    });
+    if (isJsonRpcErrorResponse(signAndPostResponse)) {
+      throw new Error(`Response must not be an error, but got '${signAndPostResponse.error.message}'`);
+    }
+    const transactionId: TransactionId = TransactionEncoder.fromJson(signAndPostResponse.result);
+    expect(transactionId).toMatch(/^[0-9A-F]+$/);
```
[the remark this answers is not in the document]
Comment (webmaster128) [reply packages/iov-core/src/jsonrpcsigningserver.spec.ts]: Because the BCP implementation of BNS guarantees to use upper case hex for transaction IDs. This property is strictly required to allow string equality on BCP's `TransactionId`s.

File: packages/iov-core/src/jsonrpcsigningserver.spec.ts [hunk 14]
```diff
@@ -0,0 +1,260 @@
...
+    if (isJsonRpcErrorResponse(response)) {
+      throw new Error(`Response must not be an error, but got '${response.error.message}'`);
+    }
+    const result = TransactionEncoder.fromJson(response.result);
+    expect(result).toEqual(jasmine.any(Array));
+    expect((result as ReadonlyArray<any>).length).toEqual(1);
+    expect(result[0]).toEqual(ganacheSecondIdentity);
+
+    server.shutdown();
+  });
```
[the remark this answers is not in the document]
Comment (webmaster128) [reply packages/iov-core/src/jsonrpcsigningserver.spec.ts]: In this test there is no Ethereum testing connection (just the one inside of the server that is disconnected during shutdown). The testing connection in the test above only exists because the chain ID in weave local test chains is generated randomly. In Ethereum tests, we use a static chain ID.

File: packages/iov-core/src/jsonrpcsigningserver.ts [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+  jsonRpcCode,
+  JsonRpcErrorResponse,
+  JsonRpcRequest,
+  JsonRpcResponse,
+  JsonRpcSuccessResponse,
+  parseJsonRpcId,
+  parseJsonRpcRequest,
+} from "@iov/jsonrpc";
+
+import { SigningServerCore } from "./signingservercore";
```
[the remark this answers is not in the document]
Comment (webmaster128) [reply packages/iov-core/src/jsonrpcsigningserver.ts]: I agree. There is no specific reason why we ended up with all lower file names.

What about CamelCase file names that match the primary symbol?

[the remark this answers is not in the document]
Comment (willclarktech) [reply packages/iov-core/src/jsonrpcsigningserver.ts]: CamelCase can cause issues, especially when combining git and OSX.

[the remark this answers is not in the document]
Comment (willclarktech) [reply packages/iov-core/src/jsonrpcsigningserver.ts]: See eg https://coderwall.com/p/mgi8ja/case-sensitive-git-in-mac-os-x-like-a-pro

File: packages/iov-core/src/transactionencoder.spec.ts (hunk 8 above)
Comment (webmaster128) [reply packages/iov-core/src/transactionencoder.spec.ts]: Well spotted. This is supposed to be _circular_, not _recursive_.

File: packages/iov-core/src/transactionencoder.ts (hunk 9 above)
Comment (webmaster128) [reply packages/iov-core/src/transactionencoder.ts]: Right. Unfortunately TypeScript's type inference is not great enough to know that. `typeof data === "object"` leads to `object | null` in this block, since `typeof null` is `object`.

File: packages/iov-core/src/transactionencoder.ts (hunk 10 above)
Comment (webmaster128) [reply packages/iov-core/src/transactionencoder.ts]: Gives this error

> Element implicitly has an 'any' type because type 'Object' has no index signature.

File: packages/iov-core/src/jsonrpcsigningserver.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (webmaster128) [reply packages/iov-core/src/jsonrpcsigningserver.ts]: The other issue I am seing is that mixed case first letters can be hard to parse or even mess up the file order. Feel free  to open an issue with a better project wide solution.

File: packages/iov-core/src/transactionencoder.ts (hunk 10 above)
Comment (webmaster128) [reply packages/iov-core/src/transactionencoder.ts]: In my experiance, the `object` type is pretty useless as it does not mean _any kind of object I don't know more about_, but basically means `{}`. I don't know the difference between `object` and `Object` but I read once that there are only very few cases where `Object` or `String` should be used as types.

File: packages/iov-core/src/jsonrpcsigningserver.ts [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+    case "getIdentities": {
+      const { reason, chainIds } = data.params;
+      if (typeof reason !== "string") {
+        throw new ParamsError("1st parameter (reason) must be a string");
+      }
+      if (!Array.isArray(chainIds)) {
+        throw new ParamsError("2nd parameter (chainIds) must be an array");
+      }
+      if (!isArrayOfStrings(chainIds)) {
+        throw new ParamsError("Found non-string element in chainIds array");
```
[the remark this answers is not in the document]
Comment (webmaster128) [reply packages/iov-core/src/jsonrpcsigningserver.ts]: The whole idea of positional indices is oudated as it turned out that JSON-RPC allows us to use names parameters. Will adapt the messages accordingly.

File: packages/iov-core/src/jsonrpcsigningserver.ts (hunk 2 above)
Comment (webmaster128) [reply packages/iov-core/src/jsonrpcsigningserver.ts]: Then we return jsonRpcCode.methodNotFound – why not! Nice catch

**Comment by webmaster128** [comment]
@willclarktech all done

File: packages/iov-core/src/jsonrpcsigningserver.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (willclarktech) [reply packages/iov-core/src/jsonrpcsigningserver.ts]: https://github.com/iov-one/iov-core/issues/927
